### DISCUSSION TITLE
How to make nested graph like android jetpack compose navigation?

### DISCUSSION THREAD

**Discussion by KangTaeJong98** [arkivanov/Decompose#408] [Q&A]
This is what I want.
https://developer.android.com/guide/navigation/design/nested-graphs

I have to separate feature module and encapsulation navigation. Just open navGraph and protect navigation.

**Comment by arkivanov** [comment]
You can add a component to [Child Stack](https://arkivanov.github.io/Decompose/navigation/stack/overview/), and that component can also use `Child Stack` and so becomes nested. This is equivalent to nested graphs.

Also see the example. [DefaultCountersComponent](https://github.com/arkivanov/Decompose/blob/master/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/DefaultCountersComponent.kt) defines its own stack. It is used in [DefaultRootComponent](https://github.com/arkivanov/Decompose/blob/master/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/root/DefaultRootComponent.kt) in outer stack. We can say that DefaultCountersComponent's graph is nested inside DefaultRootComponent.

**Reply by KangTaeJong98** [reply]
Thank you for your reply.

I think this case is little bit different. When user in startDestination on nested navigation graph press the back button, go previous navigation. But I think [DefaultCountersComponent](https://github.com/arkivanov/Decompose/blob/master/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/DefaultCountersComponent.kt) example is not.


![image](https://github.com/arkivanov/Decompose/assets/63176071/f7541588-c225-445f-8beb-b29c27f0f6f1)

**Reply by arkivanov** [reply]
In Decompose, the navigation is usually performed in the parent component holding `Child Stack`. You can pass a callback to the child component (e.g. `onFinished: () -> Unit`), and in the parent component you can manipulate `Child Stack` as needed. The parent component can bubble the event up in case the event should be handled on the upper level. So the parent component can also accept a callback, allowing the grandparent to handle the event instead. This is how nested navigation is usually performed. You can have as many nested `Child Stack`s as you want.

Does this answer your question?

**Reply by KangTaeJong98** [reply]
I understand what you said. But in android jetpack navigation, nested graph doesn't know parent graph and doesn't have callback like `onFinished: () -> Unit` just use `NavController.naviagteUp()` to go to previous route and go to parent graph. And use same `NavController`. Is there anything like this? 🤔

**Reply by arkivanov** [reply]
> But in android jetpack navigation, nested graph doesn't know parent graph

This contradicts with `NavController.naviagteUp()`, because NavController instance belongs to the parent.

In Decompose, child components are only aware of its immediate children. The child component accepts a callback, and the parent component (that is aware of the child) makes decisions how to navigate.

**Comment by KangTaeJong98** [comment]
I'm late. Thank you for your answer :)
